src/proxy/NativeMessagingProxy.cpp

```diff
@@ -0,0 +1,107 @@
+/*
+ *  Copyright (C) 2020 KeePassXC Team <team@keepassxc.org>
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "NativeMessagingProxy.h"
+#include "browser/BrowserShared.h"
+
+#include <QCoreApplication>
+#include <QtConcurrent/QtConcurrent>
+
+#include <iostream>
+
+#ifdef Q_OS_WIN
+#include <fcntl.h>
+#include <windows.h>
+#endif
+
+NativeMessagingProxy::NativeMessagingProxy()
+    : QObject()
+{
+    connect(this,
+            &NativeMessagingProxy::stdinMessage,
+            this,
+            &NativeMessagingProxy::processStandardInput,
+            Qt::QueuedConnection);
+
+    setupStandardInput();
+    setupLocalSocket();
+}
+
+void NativeMessagingProxy::setupStandardInput()
+{
+#ifdef Q_OS_WIN
+    setmode(fileno(stdin), _O_BINARY);
+    setmode(fileno(stdout), _O_BINARY);
+#endif
+
+    QtConcurrent::run([this] {
+        while (std::cin.good()) {
+            if (std::cin.peek() != EOF) {
+                unsigned int length = 0;
+                for (int i = 0; i < 4; ++i) {
+                    length |= getchar() << (i * 8);
+                }
+
+                QString msg;
+                for (unsigned int i = 0; i < length; ++i) {
+                    msg += getchar();
+                }
```

Comment:
`QStringBuilder`.

Reply:
That is an internal class and doesn't really make sense in this context. Instead I went with QString::reserve(length) and QString::append instead of the `+` operator.

Reply:
Also fine.